hollow-diff-ui/src/test/java/com/netflix/hollow/diffview/HollowHistoryUITest.java

```diff
@@ -0,0 +1,266 @@
+    public void historyUsingFwdAndRevConsumer_doubleSnapshotInRev() {
+        consumerFwd.triggerRefreshTo(3);
+        consumerRev.triggerRefreshTo(3);
+
+        historyUIServerActual = new HollowHistoryUIServer(consumerFwd, consumerRev, PORT_ACTUAL);
+
+        consumerRev.triggerRefreshTo(2);
+        consumerFwd.triggerRefreshTo(4);
+        consumerRev.triggerRefreshTo(1);
+        consumerFwd.triggerRefreshTo(5);
+
+        consumerRev.triggerRefreshTo(0);   // double snapshot in rev direction (not supported)
+    }
+
+    @Test
+    public void historyUsingFwdAndRevConsumer_backwardsCompatbileSchemaChange() throws Exception {
+        consumerFwd.triggerRefreshTo(7);    // version in whcih actor type was introduced
+        consumerRev.triggerRefreshTo(7);
+
+        historyUIServerActual = new HollowHistoryUIServer(consumerFwd, consumerRev, PORT_ACTUAL);
+
+        consumerRev.triggerRefreshTo(6);
+
+        assertNotNull(historyUIServerActual.getUI().getHistory().getHistoricalState(7).getDataAccess()
+                .getTypeDataAccess("Actor").getDataAccess());
+    }
+
+    @Test
+    public void historyUsingFwdAndRevConsumer_removeOldestState() throws Exception {
+        consumerFwd.triggerRefreshTo(3);
+        consumerRev.triggerRefreshTo(3);
+
+        historyUIServerActual = new HollowHistoryUIServer(consumerFwd, consumerRev, PORT_ACTUAL);
+
+        consumerRev.triggerRefreshTo(2);
+        consumerFwd.triggerRefreshTo(4);
+        consumerRev.triggerRefreshTo(1);
+        consumerFwd.triggerRefreshTo(5);
+
+        // drop 1 state
+        historyUIServerActual.getUI().getHistory().removeHistoricalStates(1);
+
+        // expected history is built for versions 2 through 5
+        consumerExpected = new TestHollowConsumer.Builder()
+                .withBlobRetriever(testBlobRetriever)
+                .build();
+        consumerExpected.triggerRefreshTo(2);
+        historyUIServerExpected = new HollowHistoryUIServer(consumerExpected, PORT_EXPECTED);
+        consumerExpected.triggerRefreshTo(3);
+        consumerExpected.triggerRefreshTo(4);
+        consumerExpected.triggerRefreshTo(5);
+        historyUiExpected = historyUIServerExpected.getUI();
+
+        hostUisIfPairtyCheckFails();
+    }
+
+    @Test(expected=NullPointerException.class)
+    public void historyUsingFwdAndRevConsumer_noPastVersionsAvailableAtInit()  {
+        // consumerFwd and consumerRev haven't incurred snapshot load yet
+        historyUIServerActual = new HollowHistoryUIServer(consumerFwd, consumerRev, PORT_ACTUAL);
+    }
+
+    @Test(expected=NullPointerException.class)
+    public void historyUsingFwdOnly_noPastVersionsAvailableAtInit()  {
+        // consumerFwd hasn't incurred snapshot load yet
+        historyUIServerActual = new HollowHistoryUIServer(consumerFwd, PORT_ACTUAL);
+    }
+
+    @Test(expected=UnsupportedOperationException.class)
+    public void historyUsingFwdAndRevConsumer_revConsumerMustBeInitialized() throws Exception  {
```

Comment:
How does the HistoryUIServer recover in the 2 scenarios - a) revConsumer and fwdConsumer are on different versions at init - When would that happen? b) revConsumer not initialized. Do we restart the history server with less or maybe 1 state?

Reply:
(a) The history server fails hard with an IllegalStateException at initialization when it detects that reverse consumer didnt start at the same version that the fwd consumer started with. There is also a unit test `historyUsingFwdAndRevConsumer_revAndFwdConsumersMustBeOnSameVersionAtInit` in `HollowHistoryUITest.java` that exercises this scenario. We could potentially make improvements to handle the various initialization scenarios like there is a gap in the start versions that gets stitched together,  or the fwd and rev initialize so that they will pass each other at some point. There isnt a compelling reason to handle those scenarios at the moment.

Reply:
(b) revConsumer can be initialized at any point (for e.g. after >1 historical states have been created using the fwd consumer). So if fwd and rev start at v0, at that point there are no historical states. When fwd transitions to v1 we have one historical state (historical state v1 representing the diff between v0 and v1). More fwd transition applications would lead to more historical states v2, v3, v4, etc. When the rev consumer comes along its first transition (v0 to v-1) will result in the creation of the historical state v0 (giving the same effect as if the fwd consumer had transitioned from v-1 to v0).